### PR TITLE
RAN support

### PR DESCRIPTION
**Pull request by DariushMoh**
# Description

GNodeB/UE support added, for configuration management and UERANSIM client support.

Related to [solutions/tickets#57](https://code.inmanta.com/solutions/tickets/-/issues/57)

# Reviewers list 

- [ ] @guillaume.everartsdevelp 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Correct, in line with design
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented (classes, methods and function docs, document important assumptions, link to reference materials)
- [ ] Type annotations
- [ ] Issue is in the review column
- [x] MR is assigned to first reviewer

### CODE REVIEW

File: src/nfv_test_api/config.py [hunk 1]
```diff
@@ -39,4 +41,13 @@ def get_config(config_file: Optional[str] = None, config_dict: Optional[Dict[str
             config_dict = yaml.safe_load(stream)
 
     CONFIG = Config(**config_dict)
+        
+    gnb_config_folder = pathlib.Path(CONFIG.gnb_config_folder)
+    if not gnb_config_folder.exists():
+        gnb_config_folder.mkdir()
```
Comment (edvgui) [inline src/nfv_test_api/config.py:48]:
```suggestion
    gnb_config_folder.mkdir(parents=True, exists_ok=True)
```
You can simplify this like this

File: src/nfv_test_api/v2/services/gnodeb.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+from nfv_test_api.config import Config
+
+LOGGER = logging.getLogger(__name__)
+
+class GNodeBServiceHandler:
+    def __init__(self) -> None:
+        self.processes: Dict[subprocess.Popen] = {}
+    
+    def add(self, identifier: str) -> None:
+        config_file_path = Config().gnb_config_folder+"gnb_"+identifier+".yml"
```
Comment (edvgui) [inline src/nfv_test_api/v2/services/gnodeb.py]: This is a bit asking for problems, use pathlib here as well

File: src/nfv_test_api/v2/services/gnodeb.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        command = [
+            "nr-gnb",
+            "-c",
+            config_file_path
+        ]
+
+        out_file_path = Config().gnb_log_folder+"gnb_"+identifier+".log"
+        out = open(out_file_path,"w")
```
Comment (edvgui) [inline src/nfv_test_api/v2/services/gnodeb.py]: In most cases you will actually create that file
```suggestion
        out = open(out_file_path,"w+")
```

**Review by edvgui — CHANGES_REQUESTED** [review]
I didn't make a full review yet.  Some general comments:
1. Can you use `pathlib` a bit more? For concatenating path, reading files, etc.
2. Can you update the openapi file in the repo? It would make it easier to get an overview of how the api actually looks
3. Can you extract and reuse all the logic you have to read/create files related to the gnodeb?  You build the same paths in multiple places

File: src/nfv_test_api/v2/controllers/gnodeb.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+@namespace.response(
+    code=HTTPStatus.INTERNAL_SERVER_ERROR,
+    description="An error occurred when trying to process the request, this can also be because of bad input from the user",
+)
+class GNodeBStatus(Resource):
+    def __init__(self, api=None, *args, **kwargs):
+        super().__init__(api=api, *args, **kwargs)
+        self.gnb_service = GNodeBService(Host(), gnodeb_service_handler)
+
+    @namespace.response(HTTPStatus.OK, "Found a gNodeB config with a matching nci", gnodeb_model)
```
[the remark this answers is not in the document]
Comment (edvgui) [reply src/nfv_test_api/v2/controllers/gnodeb.py:190]: image
I guess this is not what the api really returns

[the remark this answers is not in the document]
Comment (DariushMoh) [reply src/nfv_test_api/v2/controllers/gnodeb.py:190]: Indeed the start only returns status code

[the remark this answers is not in the document]
Comment (DariushMoh) [reply src/nfv_test_api/v2/controllers/gnodeb.py:190]: I will modify the header of the controller

File: src/nfv_test_api/v2/services/gnodeb.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    LOG = "log"
+
+
+def get_file_path(identifier: str, type: FileType) -> Path:
+    if type == FileType.CONFIG:
+        filename = "gnb_" + identifier + ".yml"
+        path = Path(Config().gnb_config_folder) / filename
+    elif type == FileType.LOG:
+        filename = "gnb_" + identifier + ".log"
+        path = Path(Config().gnb_log_folder) / filename
```
[the remark this answers is not in the document]
Comment (DariushMoh) [reply src/nfv_test_api/v2/services/gnodeb.py:47]: should I raise an error ?

File: src/nfv_test_api/v2/services/gnodeb.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+            universal_newlines=True,
+            stdout=out,
+            stderr=out,
+        )
+
+    def kill(self, identifier: str) -> None:
+        if identifier in self.processes:
+            self.processes[identifier].terminate()
+            self.processes[identifier].wait()
+            del self.processes[identifier]
```
[the remark this answers is not in the document]
Comment (DariushMoh) [reply src/nfv_test_api/v2/services/gnodeb.py:71]: Then nothing happens and it returns code 200. Should I raise NotFound error ?

[the remark this answers is not in the document]
Comment (edvgui) [reply src/nfv_test_api/v2/services/gnodeb.py:76]: I guess yes

**Review by edvgui — APPROVED** [review]
Looks good

**Pull request merged**